docs/source/petsc-interface.rst

```diff
@@ -280,4 +305,87 @@ before going on to solve the system as before:
            ksp.solve(b, x)
 
 
+Accessing the PETSc mesh representation
+=======================================
+
+Under the hood, Firedrake uses PETSc's DMPlex unstructured mesh
+representation. It uses a hierarchical approach, where entities
+of different dimension are put on different levels of the
+hierarchy. The single tetrahedral element shown on the left below
+may be interpreted using the graph representation on the right.
+Entities of dimension zero (vertices) are shown at the top.
+Entities of dimension one (edges) are shown on the next level down.
+Entities of dimension two (faces) are shown on the penultimate
+level and the (dimension three) element itself is on the bottom
+level. Edges in the graph indicate which entities own/are owned
+by others.
+
+.. image:: _static/dmplex.png
+   :width: 75%
+   :align: center
+
+The DMPlex associated with a given ``mesh`` may be accessed via
+its ``topology_dm`` attribute:
+
+.. code-block:: python3
+
+    plex = mesh.topology_dm
+
+All entities in a DMPlex are given a unique number. The range
+of these numbers may be deduced using the method
+``plex.getDepthStratum``, whose only argument is the entity
+dimension sought. For example, 0 for vertices, 1 for edges, etc.
+The hierarchical DMPlex structure may be traversed using other
+methods, such as ``plex.getCone``, ``plex.getSupport`` and
+``plex.getTransitiveClosure``. See the `Firedrake DMPlex paper`_
+for details.
+
+If vertex coordinate information is to be accessed from the
+DMPlex then we must first establish a mapping between its
+numbering and the coordinates in the Firedrake mesh. This is done
+by establishing a 'section'. A section can be thought of as a field
+defined upon the mesh - in this case, the coordinate field.
+For a $d$-dimensional mesh, we seek to establish offsets to recover
+the a $d$-tuple of coordinates. That is, we seek $d$ vertex-wise
+values and no values for entities of higher dimension.
+
+In 2D, for example, this corresponds to the array
+
+.. math::
+
+   (d, 0, 0)
+
+Accordingly, set
+
+.. code-block:: python3
+
+    dim = mesh.topological_dimension()
+    entity_dofs = np.zeros(dim+1, dtype=np.int32)
+    entity_dofs[0] = mesh.geometrical_dimension()
+
+We then use Firedrake's helper function for creating a PETSc
+section to establish the mapping:
+
+.. code-block:: python3
+
+    from firedrake.cython.dmcommon import create_section
+
+    coord_section = create_section(mesh, entity_dofs)
+    plex = mesh.topology_dm
+    plex_coords = plex.getCoordinateDM()
+    plex_coords.setDefaultSection(coord_section)
+    coords_local = plex_coords.createLocalVec()
+    coords_local.array[:] = np.reshape(mesh.coordinates.dat.data_ro_with_halos, coords_local.array.shape)
+    plex.setCoordinatesLocal(coords_local)
```

Comment:
FWIW, if the coordinate field is piecewise linear, this is what is needed to migrate the firedrake coordinates to PETSc (@ReubenHill) (c.f. #2185)